File: source/jugglinglab/renderer/JLVector.java
// JLVector.java
//
// Copyright 2002-2022 Jack Boyce and the Juggling Lab contributors

package jugglinglab.renderer;


public class JLVector {
    public double x;
    public double y;
    public double z;


    public JLVector() {
    }

    public JLVector(double xpos, double ypos, double zpos) {
        x = xpos;
        y = ypos;
        z = zpos;
    }

    public double length() {
        return Math.sqrt(x*x + y*y + z*z);
    }

    public JLVector transform(JLMatrix m) {
        double newx = x * m.m00 + y * m.m01 + z * m.m02 + m.m03;
        double newy = x * m.m10 + y * m.m11 + z * m.m12 + m.m13;
        double newz = x * m.m20 + y * m.m21 + z * m.m22 + m.m23;
        return new JLVector(newx, newy, newz);
    }

    public static JLVector add(JLVector a, JLVector b) {
        return new JLVector(a.x + b.x, a.y + b.y, a.z + b.z);
    }

    public static JLVector sub(JLVector a, JLVector b) {
        return new JLVector(a.x - b.x, a.y - b.y, a.z - b.z);
    }

    public static JLVector scale(double f, JLVector a) {
        return new JLVector(f * a.x, f * a.y, f * a.z);
    }
}

File: src/main/kotlin/jugglinglab/renderer/JLVector.kt
//
// JLVector.kt
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.renderer

import jugglinglab.util.Coordinate
import kotlin.math.sqrt

class JLVector {
    var x: Double = 0.0
    var y: Double = 0.0
    var z: Double = 0.0

    constructor()

    constructor(xpos: Double, ypos: Double, zpos: Double) {
        x = xpos
        y = ypos
        z = zpos
    }

    val length: Double
        get() = sqrt(x * x + y * y + z * z)

    fun transform(m: JLMatrix): JLVector {
        val newx = x * m.m00 + y * m.m01 + z * m.m02 + m.m03
        val newy = x * m.m10 + y * m.m11 + z * m.m12 + m.m13
        val newz = x * m.m20 + y * m.m21 + z * m.m22 + m.m23
        return JLVector(newx, newy, newz)
    }

    companion object {
        fun add(a: JLVector, b: JLVector): JLVector {
            return JLVector(a.x + b.x, a.y + b.y, a.z + b.z)
        }

        fun sub(a: JLVector, b: JLVector): JLVector {
            return JLVector(a.x - b.x, a.y - b.y, a.z - b.z)
        }

        fun scale(f: Double, a: JLVector): JLVector {
            return JLVector(f * a.x, f * a.y, f * a.z)
        }

        fun fromCoordinate(c: Coordinate, result: JLVector): JLVector {
            result.x = c.x
            result.y = c.z
            result.z = c.y
            return result
        }
    }
}
